docs/design/run.md

```diff
@@ -0,0 +1,268 @@
+We should be able to parallelize running the actual command, while preserving a
+good console output.
+The run command should be able to work in the working copy. 
+There should exist some way to signal hard failure. 
+The command should build enough infrastructure for `jj test`, `jj fix` and 
+`jj format`.
+The main goal is to be good enough, as we can always expand the functionality 
+in the future.
+
+### Non-Goals:
+
+While we should build a base for `jj test`, `jj format` and `jj fix`, we 
+shouldn't mash their use-cases into `jj run`.
+The command shouldn't be too smart, as too many assumptions about workflows 
+makes the command confusing for users. 
+The smart caching of outputs, as user input commands can be unpredictable.
+Fine grained user facing configuration, as it's unwarranted complexity.
+A `fix` subcommand as it cuts too much design space.
+
+## Use-Cases of jj run:
+
+**Linting and Formatting:**
+- `jj run 'pre-commit run' -r $revset`
+- `jj run 'cargo clippy' -r $revset`
+- `jj run 'cargo +nightly fmt'`
+**Large scale changess across repositories, local and remote:**
+- `jj run 'sed s/some/test' -r 'draft() & ~remote_branches(exact:"origin")'`
+- `jj run '$rewrite-tool' -r '$revset'`
+**Build systems:**
+- `jj run 'bazel build //some/target:somewhere'`
+- `jj run 'ninja check-lld
+
+Some of these use-cases should get a specialized command, as this allows 
+further optimization. A command could be `jj format`, which runs a list of 
+formatters over a subset of a file in a revision. Another command could be 
+`jj fix`, which runsa command like `rustfmt --fix` or `cargo clippy --fix` over
+a subset of a file in a revision.
+
+## Design:
+
+### Base Design: 
+
+All the work will be done in the `.jj/` directory. This allows us to hide all 
+complexity from the users, while preserving the user's current workspace.
+
+We will copy the approach from git-branchless's `git test` of creating a 
+temporary working copy for each parallel command. The working copies will be 
+reused between `jj run` invocations. They will also be reused within `jj run` 
+invocation if there are more commits to run on than there are parallel jobs.
+
+We will leave ignored files in the temporary directory between runs. That 
+enables incremental builds (e.g by letting cargo reuse its `target/` directory).
+However, it also means that runs potentially become less reproducible. We will 
+provide a flag for removing ignored files from the temporary working copies to
+address that. 
+
+Another problem with leaving ignored files in the temporary directories is that
+they take up space. That is especially problematic in the case of cargo (the 
+`target/` directory often takes up tens of GBs). The same flag for cleaning up
+ignored files can be used to address that. We may want to also have a flag for 
+cleaning up temporary working copies *after* running the command. 
+
+An early version of the command will directly use [Treestate][treestate] to 
```

Comment:
The kind of link you use here doesn't seem to work in Github. I'm unsure whether it'd work with `mkdocs serve`.